text/0000-alternate-registries.md

```diff
@@ -0,0 +1,120 @@
+* Add an optional 'registry' field to be specified in the package definition
+
+Both of the registry entries take a value of a crates.io git repository. If one is not
+provided then the default crates.io repository is assumed, this ensures that it is
+back compatible with all current crates.
+
+Below is an example of the change that we are proposing to make:
+
+```
+[package]
+name = "registry-test"
+version = "0.1.0"
+authors = ["Christopher Swindle <christopher.swindle@metaswitch.com>"]
+registry = "https://github.com/my_awesome_fork/crates.io-index"
+
+[dependencies]
+libc = { version = "*", registry = "https://github.com/my_awesome_fork/crates.io-index" }
+serde_json = "1.0.1"
+```
+
+## Changes for alternative registries for dependencies
+This boils down to a very simple change, where we previously setup the crate source for the
+crates.io registry, we now just need to check if a registry is provided, if it has the crate
+source is created using the registry URL, otherwise the crates.io server is used.
+
+## Blocking requests to push to a registry
+There are two parts to this, the first is a change to Cargo which checks if the registry
+provided in the registry matches the host for the publish, if it does not it gets rejected.
+The second part is a change to crates.io which will just reject the request to publish the
+crate if the configured repository on the crates.io server does not match the registry
+specified in the package or dependencies.
+
+## Allow publishing when referencing external dependencies
+We still want to support private crates having dependencies on the public crates.io server,
+so we propose relaxing a check which ensures that the source for a dependency matches the
+registry. We propose that this only performs the check only if the dependency is not the
+default registry, thus allowing private crates to reference public crates on crates.io.
+
+## Making it easier for users using an alternate crates.io registry
+When a user selects a specific crate the Cargo.toml fragment would be updated to include the
+registry URL, thus allowing users to easily copy and paste into their projects Cargo.toml
+file.
+
+# How We Teach This
+[how-we-teach-this]: #how-we-teach-this
+
+The term alternative registry would seem the most appropriate to describe this feature.
+
+In the first instance I think that the Cargo.toml format documentation is sufficient to
+provide access to the feature. However, in time once more of the pieces fall into place
+it would be useful having a guide on how to setup/administer a crates.io server in an
+enterprise setting (similar to the initial mirroring documentation).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Currently this design requires that when you want to push to the private crates.io server
+you need to override the host and token, it would be possible to update cargo to support
+multiple registries tokens which can be used to login.
```

Comment:
I agree and that is something which I would like as well, but I wanted to try and get something in place, which can then be built on to add support for things like this.